### PR TITLE
chore: Bump numpy support

### LINKED ISSUE #407: Bump numpy version
Looks like numpy 2.0 is just around the corner: https://github.com/numpy/numpy/issues/24300. But in the meantime we should still update the numpy version currently used, since we are still using `1.7` deprecated api:
- Remove the `#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION` and fix the api
- Numpy has minimum version of `3.9`, maybe we should bump it as well, although `pip install` should pick up the appropriate version automatically

### PR DESCRIPTION
**Pull request by LecrisUT**
- Bump numpy support
- Refactor python CI tests

Closes #407

### CODE REVIEW

**Review by lan496 — APPROVED** [review]

File: python/_spglib.c
Comment (LecrisUT) [inline python/_spglib.c:1]: Opening a comment on a random file just to have somewhere to discuss. @rgommers regarding the `build-system.require`

If you check the tests in this project, you will see that we have tight collaboration with downstream packaging, and we do our best to check the packaging. I would prefer not to introduce dependency locking unless well documented. Looking at who would be affected by such a change:
- PyPI wheel: we have CD for releasing to PyPI and because numpy 2.0 has been officially released, that version will automatically be used on the next release
- Packagers: the build with no isolation,  and adding a version lock would just make more work for them to patch it out for no benefit on their side
- Users (e.g. installing via `pip install git+https://...`): with significant effort, they can choose which version of numpy to use for build. But by default the latest version should be used

Are there aspects that we should consider?

File: python/_spglib.c
Comment (LecrisUT) [inline python/_spglib.c:37]: I chose 1.20 simply because that's the version on the latest RHEL9 (lower than Ubuntu 24.04). Any comments on this? I think adding a `build-system.requires` linked to this would be advisable?

**Review by LecrisUT — COMMENTED** [review]
Should we include any tests for both numpy 1.x and 2.x for runtime, or is it not worth the complexity?

File: python/_spglib.c
Comment (rgommers) [reply python/_spglib.c:1]: Hi @LecrisUT, good questions. I'm very much aware of downstream packaging needs (and trying to make Python packaging better in that respect ... slow process).

> Packagers: the build with no isolation, and adding a version lock would just make more work for them to patch it out for no benefit on their side

All packagers indeed disable build isolation. In that case there is nothing to patch though, the effect of disabling build isolation is that `build-system.requires` is completely ignored. So this isn't very relevant to them either way. And adding the code comment I suggested in https://github.com/spglib/spglib/issues/407#issuecomment-2185909030 clears up any potential confusion.

> Users (e.g. installing via `pip install git+https://...`): with significant effort, they can choose which version of numpy to use for build. But by default the latest version should be used

This is the most fragile place. If users installs `spglib` directly, then yes, build isolation will pull in latest numpy. However, if they build spglib as a dependency (e.g., `pip install pkg-with-spglib-dep`), then this can go wrong; you'll get the `numpy` version that's already put in the build env by the other package. So if that package has a constraint `numpy<2.0`, you silently get builds that are not going to work. And worse, `pip` will then cache the bad `spglib` wheel, so after that you keep on getting the wrong version in other envs too. This can be very confusing for users.

* * *

So my assessment of your three points:
- For your own releases it's the same - but explicit is slightly better than implicit, so adding the constraint is slightly nicer
- For downstream packagers it doesn't matter, and they're used to the constraint. I recommend copying exactly what SciPy does here.
- For users, you really do need the constraint to avoid some of the sharp edges of Python packaging.

File: python/_spglib.c
Comment (rgommers) [reply python/_spglib.c:37]: This seems fine to me.

**Comment by rgommers** [comment]
> Should we include any tests for both numpy 1.x and 2.x for runtime, or is it not worth the complexity?

It's generally nice to test with the lowest and highest supported version of a dependency, especially when binary compatibility and bc-breaking changes are involved. So it sounds useful to me - however I don't know much about `spglib`; in case you use very little of the numpy API then it may not be worth the effort.

File: python/_spglib.c
Comment (LecrisUT) [reply python/_spglib.c:1]: > All packagers indeed disable build isolation. In that case there is nothing to patch though, the effect of disabling build isolation is that build-system.requires is completely ignored. So this isn't very relevant to them either way.

The `build-system.requires` is still used, like in:
https://github.com/spglib/spglib/blob/9acbe6dcd8ceedbec76426b6d107c46a6c9b7484/.distro/spglib.spec#L60-L61
If you look at the spec file there, `python-scikit-build-core` (or equivalents) are not defined anywhere, instead they are parsed from the `pyproject.toml`. The patching that is needed for packagers is to delete the `>=2.0` there.

> So if that package has a constraint `numpy<2.0`, you silently get builds that are not going to work.

Good point, let me meditate on this for a while. I will raise this issue on the Fedora Python forum, and get back on the opinions there.

At the same time, if the dependency has a `numpy<2.0` in the `build-system.requires`, they would also have to have an equivalent in the `dependencies`. If we add the `numpy>=2.0` it would just make it unbuildable altogether, even though those dependencies should still be suported. Of course if wheels are allowed, than this issue is void. I guess numpy 2.x does not provide useful information for the user to know when numpy 1.x was used and it needs to be rebuilt?

A small concern I have is with supporting RHEL10 and Ubuntu 24.04 which just missed the numpy 2.x change, but `scikit-build-core` was included on time, so it should be packageable there (could not update for a few versions because of the build-system change)

Comment (rgommers) [reply python/_spglib.c:1]: > If you look at the spec file there, `python-scikit-build-core` (or equivalents) are not defined anywhere, instead they are parsed from the `pyproject.toml`. The patching that is needed for packagers is to delete the `>=2.0` there.

Okay, thanks. I have to admit I've never seen such a file as part of Python package, nor can I tell who actually uses it. What I meant it's not used by Python packaging; this is custom code, aimed at one or a handful of distros I guess?

Comment (LecrisUT) [reply python/_spglib.c:1]: The spec file is basically the instructions for how a RPM package is built, which we are testing and synchronizing with Fedora right now. This is basically an integration with downstream packagers in order to guarantee compatibility

Comment (rgommers) [reply python/_spglib.c:1]: > I guess numpy 2.x does not provide useful information for the user to know when numpy 1.x was used and it needs to be rebuilt?

I'm not sure I understand this question. A wheel built against 2.x is compatible with all 1.x and 2.x versions, so nothing needs to be rebuilt. If 1.20.0 is your lowest-supported numpy version, then this is the default recommendation for metadata for correct ABI handling and the widest-possible runtime version range support:
```toml
[build-system]
requires = ["numpy>=2.0.0,<3"]

[project]
dependencies = ["numpy>=1.20,<3]",
```

Comment (LecrisUT) [reply python/_spglib.c:1]: Thinking about making sure that the wheel always has the supported ABI compatible versions, it kinda makes sense. But then again, you can simply run `pip install --upgrade numpy>3`. I keep ping-ponging on this aspect.


> > I guess numpy 2.x does not provide useful information for the user to know when numpy 1.x was used and it needs to be rebuilt?
> 
> I'm not sure I understand this question

What I was considering about this idea is if numpy would have some functionality built-in or provide a function that can be used at the python level to check if a package is ABI compatible or not, and give appropriate error message that Package X needs to be rebuilt for the runtime-numpy version used or downgrade the runtime-numpy version.

Comment (rgommers) [reply python/_spglib.c:1]: You'll get a message, or one of several possible ones about binary compat. For Cython code, when you compile with 1.x and at runtime use 2.x, you'll see:
```
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```
It's hinting at the problem, but it's not explicit enough (Cython can only see mismatching structs). If you only use pure C/C++ extensions, you may get a clearer message stating "ABI mismatch ... (version numbers)" from NumPy itself.

> What I was considering about this idea is if numpy would have some functionality built-in or provide a function that can be used at the python level to check if a package is ABI compatible or not,

That you could build in principle, but there's nothing out of the box that does it right now. At the C level there are API and ABI version numbers separate from the package version, but at the Python level the major version number of numpy itself is what you'd use. When ABI number at the C level changes, the major version number has to be bumped. So at runtime you can look at `np.__version__[0]`.

> But then again, you can simply run `pip install --upgrade numpy>3`. I keep ping-ponging on this aspect.

This passes on the problem to users, so I recommend not doing that. Also, the `<3` is correct no matter what you do with the rest, since it's certain that 3.0 (if that ever happens; it may not) will be another ABI break.

In the end the problem is that metadata for Python packaging and for distros fundamentally are not the same - this is a limitation of Python packaging. The best solution here I think is to drop `%generate_buildrequires` and simply add `BuildRequires: ...` for your build dependencies instead. You only have two build dependencies in `pyproject.toml`, so this can't be considered a maintenance burden really - you can just add a comment above these things in both metadata files that explains what is going on with them, and that if you update one of them, to also check the other.

Comment (LecrisUT) [reply python/_spglib.c:1]: > The best solution here I think is to drop `%generate_buildrequires` and simply add `BuildRequires: ...`

`%generate_buildrequires` handles the `[test]` optional dependencies as well, so not really an option. I have added a patch to be shipped with for packagers, but I will still wait for some feedback from fellow packagers before resolving this thread.

> This passes on the problem to users, so I recommend not doing that

I was considering more along the lines, if we are sure that adding the `dependencies` restriction would guarantee the user doesn't shoot themselves in the foot when doing `pip --upgrade numpy >= 3.0`. Afaik, that is not the case since pip ignores any installed package's dependencies.

So it seems this would mostly benefit the first installation and it will disallow dependencies of packages that have `build-system.requires` with `numpy<2`

> If you only use pure C/C++ extensions, you may get a clearer message stating "ABI mismatch ... (version numbers)" from NumPy itself.

This is reassuring.

> At the C level there are API and ABI version numbers separate from the package version

Need to confirm something, when does the ABI incompatibility kick in? Isn't it at the library linkage time, such that even if I added a check in `PyInit`, that would not be executed because of incompatible symbols?

Comment (rgommers) [reply python/_spglib.c:1]: It's not a link-time thing, because there's no direct dependency on a `numpy_xxx.so`. It kicks in at the place where you access the numpy C API, probably through `import_array()`, `import_ufunc()`, or a similar numpy C API function. I can't find so quickly where and how in your code base you actually use NumPy headers.

Comment (LecrisUT) [reply python/_spglib.c:1]: > It's not a link-time thing, because there's no direct dependency on a `numpy_xxx.so`. It kicks in at the place where you access the numpy C API

Oh that's useful to consider

> probably through `import_array()`, `import_ufunc()`, or a similar numpy C API function.

On this part, I believe we should migrate to more modern numpy interface. Review on this would be appreciated.

> I can't find so quickly where and how in your code base you actually use NumPy headers.

Example:
https://github.com/spglib/spglib/blob/9acbe6dcd8ceedbec76426b6d107c46a6c9b7484/python/_spglib.c#L1471-L1487

Comment (rgommers) [reply python/_spglib.c:1]: > On this part, I believe we should migrate to more modern numpy interface. Review on this would be appreciated.

It looks like you're using NumPy C API functionality in that file without either initializing with `import_array()` or defining `NO_IMPORT_ARRAY`/`PY_ARRAY_UNIQUE_SYMBOL`. 

I think the docs at https://numpy.org/devdocs/reference/c-api/array.html#including-and-importing-the-c-api should explain what is needed here.

Comment (LecrisUT) [reply python/_spglib.c:1]: Thanks for the links and sorry for being slow. I've read a bit on what this does and afaiu, the only thing missing in the bindings was the `import_array` (maybe some better error catching, but that's another issue). If you see anything else that seems suspicious, let me know.

Btw, all of this is done under-the-hood with something like `pybind11`? Considering migrating to that or `nanobind`.

Comment (rgommers) [reply python/_spglib.c:1]: Great! And yes, if you're comfortable with C++ then `pybind11` or `nanobind` are great. They offer a little more than only binding generation, they also have a bit of a C++ API for array data structures, so when you're writing something from scratch you may have a slightly higher level API to work with than `PyArray_*`.

**Comment by LecrisUT** [comment]
Ok, all :+1: from my side on this issue. Fedora packaging seems to go ok, although rawhide is not on numpy 2.0 yet, and the new test where we downgrade the build numpy to 1.x seems to work fine. @rgommers just one final confirmation, building with `numpy 1.x` and running with `2.x` is unsupported?

Other than that @lan496 I leave it to you to make a final review and merge

**Comment by rgommers** [comment]
> @rgommers just one final confirmation, building with `numpy 1.x` and running with `2.x` is unsupported?

Yes, that definitely will not work.

**Review by lan496 — APPROVED** [review]
Note: This PR also bumps up the required python version to 3.9

**Pull request merged**
